Remove "Ask Brave for suggestions" from context menu spellcheck section

mkarolin opened the pull request:
Hides RenderViewContextMenu::AddSpellCheckServiceItem method with one in BraveRenderViewContextMenu that does not add Ask Brave menu item.

Fixes brave/brave-browser#1623

## Submitter Checklist:

- [ ] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [x] Windows
  - [x] macOS
  - [ ] Linux
- Verified that these changes pass automated tests (`npm test brave_unit_tests && npm test brave_browser_tests`) on
  - [x] Windows
  - [x] macOS
  - [ ] Linux
- [ ] Ran `git rebase master` (if needed).
- [ ] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [x] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone

## Automated Test Plan:

On all platforms:
`npm run test -- brave_browser_tests --filter=BraveSpellingMenuObserverTest.*`

On Windows and Linux:
`npm run test -- brave_browser_tests --filter=BraveSpellingOptionsSubMenuObserverTest.*`

## Test Plan:

1. [Windows & Linux] Verify that Spellcheck sub-menu doesn't contain "Ask Brave for suggestions":
- Start the browser and open brave://settings/languages,
- Expand "Spell check" section and make sure that at least one language is turned on for spell checking,
- In a new tab navigate to google.com,
- Right-click in the search box on the site and expand Spellcheck sub-menu,
- Verify that the sub-menu doesn't contain a menu item "Ask Brave for suggestions".

2. [Windows & Linux] Verify that Spellcheck sub-menu doesn't end with a separator when spell checking is turned off for all languages:
- Start the browser and open brave://settings/languages,
- Expand "Spell check" section and turn off spell checking for all languages,
- In a new tab navigate to google.com,
- Right-click in the search box on the site and expand Spellcheck sub-menu,
- Verify that the sub-menu doesn't end with a separator.

3. [All Platforms] Verify that spelling suggestions don't show "Ask Brave for suggestions":
- Start the browser and open brave://settings/languages,
- Expand "Spell check" section and make sure that at least one language is turned on for spell checking,
- In a new tab navigate to dictionary.com,
- Right-click in the search box on the site and expand Spellcheck sub-menu, make sure that "Check for spelling of text fields" menu item is checked,
- Type a misspelled word into the search box on the site (e.g. correctz),
- Right click on the misspelled word,
- Verify that "Ask Brave for suggestions" doesn't show in the menu (after "Add to dictionary").

## Reviewer Checklist:

- [x] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [x] Adequate test coverage exists to prevent regressions 
- [x] Verify test plan is specified in PR before merging to source

bsclifton commented:
@mkarolin would it be possible to make a browser test for this? I'm not 100% sure context menu items can be tested, but it would be cool to check 😄. If it was possible, a great test would be to ensure this item never shows up when right clicking a text field

bsclifton commented:
Change worked as expected- however, it leaves an awkward separator above where the item used to be:
<img width="606" alt="screen shot 2018-10-18 at 2 20 40 pm" src="https://user-images.githubusercontent.com/4733304/47185166-84a9df80-d2e1-11e8-866a-7ac8d48d1e74.png">

It might be better to look at the caller for this, which seems to be:
`src/chrome/browser/renderer_context_menu/spelling_menu_observer.cc`
(see `InitMenu()`)

While we're there, you might check other places that use `integrate_spelling_service_` member variable (which I think is tied to the Profile preference `spellcheck.use_spelling_service`, which is set to false by default)

mkarolin commented:
Oh that's interesting, I always have another item there:

![image](https://user-images.githubusercontent.com/41635752/47192974-85149b80-d31e-11e8-901a-ba039e8e6cd1.png)

I'll take a closer look to see what appears when. Thanks for trying it out!

mkarolin commented:
Updated. There were 2 places where "Ask Brave for suggestions" could show up:
1. In the Spellcheck sub-menu,
2. In the menu if spellcheck is enabled for text fields and the menu is invoked on a misspelled word.
Updated test plan to check for both cases.
Added browser tests to verify these cases and to verify the Spellcheck sub-menu doesn't end with a separator.

bsclifton commented on browser/renderer_context_menu/brave_spelling_options_submenu_observer.cc:
```diff
@@ -0,0 +1,68 @@
+  DCHECK(submenu_model_.GetItemCount());
+
+  // Special accommodations for gtest.
+  if (gtest_mode_ != GTEST_MODE_DISABLED) {
```
Just a thought for the future (I'm curious too) - you might check if unit/browser tests expose a `#define` that you can check for with preprocessors? ex: you could wrap this section (and potentially the definition/setting of `gtest_mode_`) with `#ifdef`/`#endif` guards

mkarolin replied:
I was looking for something like this as well but didn't see any additional definitions in the tests build in our build setup. gtest itself doesn't seem to add any (https://stackoverflow.com/questions/6077833/how-to-check-if-google-test-is-running-in-my-code). Ideally, I wouldn't want to have a different code path for tests in the first place, but in this case it would mean creating a more complex mock object (the current one was taken from Chromium and slightly modified).

bsclifton commented on browser/renderer_context_menu/brave_spelling_options_submenu_observer.cc:
```diff
@@ -0,0 +1,68 @@
+  int index = submenu_model_.GetItemCount() - 1;
+  if (index >= 0 &&
+      submenu_model_.GetTypeAt(index) == ui::MenuModel::TYPE_SEPARATOR)
+    submenu_model_.RemoveItemAt(index);
```
nitpick: when a comparison is multiple lines (or has weird spacing, enclosing the block within `{` and `}` can make it more readable (clearly capturing the start/end)

bsclifton commented:
Basically ready for merge - just going to verify tests work great on macOS, since this option isn't available there (build in progress...)

bsclifton commented:
Getting some compile errors (likely need platform preprocessor guards around this):
```
../../brave/browser/renderer_context_menu/brave_spelling_options_submenu_observer.cc:52:35: error: use of undeclared identifier 'IDS_CONTENT_CONTEXT_SPELLCHECK_MENU'
        l10n_util::GetStringUTF16(IDS_CONTENT_CONTEXT_SPELLCHECK_MENU),
                                  ^
../../brave/browser/renderer_context_menu/brave_spelling_options_submenu_observer.cc:62:35: error: use of undeclared identifier 'IDS_CONTENT_CONTEXT_SPELLCHECK_MENU'
        l10n_util::GetStringUTF16(IDS_CONTENT_CONTEXT_SPELLCHECK_MENU));
                                  ^
2 errors generated.
```

mkarolin commented:
Updated with ifdefs for Mac. One curious thing: in the original render_view_context_menu.cc code the header inclusion is predicated on `#if !BUILDFLAG(USE_BROWSER_SPELLCHECKER)` whereas the use of SpellingOptionsSubMenuObserver is controlled by `if !defined (OS_MACOSX)`. I kept it the same in the override.

bsclifton commented:
Built great on macOS and all unit/browser tests passed; waiting for Windows to finish...

bsclifton commented:
master d743327cfdd7196d601e96578b66357693588063
0.57.x 5d8323aef9df179d12e448b66b31fb5111350d70
0.56.x d674e0ba6ace017b5cc51b457f7aaab39a1b8c2a
